// Copyright 2020 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style
// license that can be found in the LICENSE file or at
// https://developers.google.com/open-source/licenses/bsd

// This file defines protobufs for users and related business
// objects, e.g., users, user preferences.

syntax = "proto3";

package monorail.v1;

import "google_proto/google/api/resource.proto";
import "google_proto/google/api/field_behavior.proto";

// User represents a user of the Monorail site.
// Next available tag: 4
message User {
  option (google.api.resource) = {
    type: "api.crbug.com/User"
    pattern: "users/{user_id}"
  };
  // Resource name of the user.
  string name = 1;
  // Obscured or un-obscured user email or name to show other users using the site.
  string display_name = 2 [ (google.api.field_behavior) = OUTPUT_ONLY ];
  // User-written indication of their availability or working hours.
  string availability_message = 3;
}


// UserSettings represents preferences and account settings of a User.
// Next available tag: 4
message UserSettings {
  option (google.api.resource) = {
    type: "api.crbug.com/UserSettings"
    pattern: "usersettings/{user_id}"
  };
  // Resource name of the user that has these settings.
  string name = 1 [ (google.api.resource_reference) = {type: "api.crbug.com/UserSettings"} ];
  // The global site role for a User.
  // Next available tag: 3
  enum SiteRole {
    // Default value. This value is unused.
    SITE_ROLE_UNSPECIFIED = 0;
    // Normal site user with no special site-wide extra permissions.
    NORMAL = 1;
    // Site-wide admin role.
    ADMIN = 2;
  }
  SiteRole site_role = 2;
  // Resource name of a linked primary User.
  string linked_primary_user = 3 [
      (google.api.resource_reference) = {type: "api.crbug.com/User"},
      (google.api.field_behavior) = OUTPUT_ONLY ];
}

// Defines saved queries that belong to a user
//
// Next available tag: 5
message UserSavedQuery {
  option (google.api.resource) = {
    type: "api.crbug.com/UserSavedQuery"
    pattern: "users/{user_id}/savedQueries/{savedQuery}"
  };

  // Resource name of this saved query.
  string name = 1;
  // Display name of this saved query, ie 'open issues'.
  string display_name = 2;
  // Search term of this saved query.
  string query = 3;
  // Subscription mode of this saved query
  // Next available tag: 3
  enum SubscriptionMode {
    // Default API value. This value is unused.
    SUBSCRIPTION_MODE_UNSPECIFIED = 0;
    // Do not subscribe to notifications.
    NO_NOTIFICATION = 1;
    // Subscribe to notifications.
    IMMEDIATE_NOTIFICATION = 2;
  }
  SubscriptionMode subscription_mode = 4;
}